File: collisions/simple_collisions.py
import time

# Simulating the collision between both particles, assuming the coefficient of restitution is 1
# Input vec: a vector representing v_a, v_b

# Formula for restitution when e = 1: v_b + -v_a = u_a - u_b
# Formula for conservation of momentum: m_a*u_a + m_b*u_b = m_a*v_a + m_b*v_b

def col_p(vec, mass_vector):
    u_a = vec[0]
    u_b = vec[1]
    m_a = mass_vector[0]
    m_b = mass_vector[1]
    
    # Found by substituting the formula for restitution into the formula for conservation of momentum to cancel out v_b, then rearranging for v_a
    final_velocity_a = (m_a*u_a + 2*m_b*u_b - m_b*u_a)/(m_a+m_b)
    # Same as above, but cancel out v_a and rearrange for v_b
    final_velocity_b = (m_b*u_b + 2*m_a*u_a - m_a*u_b)/(m_a+m_b)

    return [final_velocity_a, final_velocity_b]

# Simulating the collision between a particle and the magical wall, again assuming the coefficient of restitution is 1. particle is either 0 (a) or 1 (b).
def col_w(vec, particle):
    if particle == 0:
        return [-vec[0], vec[1]]
    else:
        return [vec[0], -vec[1]]
    # That was easy :)

# Check if no more collisions will occur
def check_velocity(vec):
    if (vec[0] >= 0) and (vec[1] >= 0):
        if vec[0] > vec[1]:
            return True
    
    return False

def main(n, verbose=True):
    n = n # Number of significant figures of pi to calculate

    m_a = 100**(n-1) # Mass of particle a
    m_b = 1 # Mass of particle b

    mass_vector = [m_a, m_b]

    initial_velocity_a = -1
    initial_velocity_b = 0
    vec = [initial_velocity_a, initial_velocity_b]
    collisions = 0
    while True:
        # Collision between particles
        vec = col_p(vec, mass_vector)
        collisions += 1
        if check_velocity(vec):
            break

        # Collision between wall and particle
        vec = col_w(vec, 1) # Only particle b can collide with wall as it is between the wall and particle a
        collisions += 1
        if check_velocity(vec):
            break
    
    if verbose:
        print("The mass of Particle A is: " + str(m_a) + "kg")
        print("The mass of Particle B is: " + str(m_b) + "kg")

        print("The initial velocity of Particle A was " + str(initial_velocity_a) + "m/s")
        print("The initial velocity of Particle B was " + str(initial_velocity_b) + "m/s")

        print("The final velocity of Particle A was " + str(vec[0]) + "m/s")
        print("The final velocity of Particle B was " + str(vec[1]) + "m/s")

        print("The total number of collisions was " + str(collisions))
        print("This means that Pi to " + str(n) + " significant figures is " + str(collisions/(10**(n-1))))

    return collisions

main(5)

# Timing function
# Time taken increases by roughly a factor of 10 for each extra significant figure.
for n in range(1, 10):

    start = time.time()
    pi = main(n, verbose=False)
    end = time.time()
    print("Calculating pi to " + str(n) + " significant figures gave result " + str(pi/(10**(n-1))) + " and took " + str(end-start) + " seconds.")

File: collisions/simple_collisions.rs

fn col_p(u_a: f64, u_b: f64, m_a_integer: i128, m_b_integer: i128) -> (f64, f64) {

    let m_a = m_a_integer as f64;
    let m_b = m_b_integer as f64;

    // Found by substituting the formula for restitution into the formula for conservation of momentum to cancel out v_b, then rearranging for v_a
    let final_velocity_a = (m_a*u_a + 2.0*m_b*u_b - m_b*u_a)/(m_a+m_b);
    // Same as above, but cancel out v_a and rearrange for v_b
    let final_velocity_b = (m_b*u_b + 2.0*m_a*u_a - m_a*u_b)/(m_a+m_b);

    return (final_velocity_a, final_velocity_b)
}

fn col_w(v_a: f64, v_b: f64, particle: bool) -> (f64, f64) {
    if particle {
        return (-v_a, v_b);
    } else {
        return (v_a, -v_a);
    }
}

fn check_velocity(v_a: f64, v_b: f64) -> bool {
    if v_a >= 0.0 && v_b >= 0.0 && v_a >= v_b {
        return true;
    } else {
        return false;
    }
}

fn main() {
    let n = 10_u32;
    let m_a: i128 = 100_i128.pow(n - 1);
    let m_b = 1_i128;

    let initial_velocity_a = -1_f64;
    let initial_velocity_b = 0_f64;

    let mut collisions: u128 = 0;
    
    println!("The initial velocity of particle A is {} m/s", initial_velocity_a);
    println!("The initial velocity of particle B is {} m/s", initial_velocity_b);
    println!("The initial mass of particle A is {} kg", m_a);
    println!("The initial mass of particle B is {} kg", m_b);

    let mut v_a = initial_velocity_a;
    let mut v_b = initial_velocity_b;

    loop {
        (v_a, v_b) = col_p(v_a, v_b, m_a, m_b);
        collisions += 1;
        if check_velocity(v_a, v_b) {
            break;
        }

        (v_a, v_b) = col_w(v_a, v_b, true);
        collisions += 1;
        if check_velocity(v_a, v_b) {
            break;
        }
    }

    let pi: f64 = (collisions as f64)/(10_u64.pow(n - 1)) as f64;
    println!("The final velocity of particle A is {} m/s", v_a);
    println!("The final velocity of particle B is {} m/s", v_b);
    println!("There were {}", collisions);
    println!("Pi to {} significant figures is {}", n, pi);
}
